api_core/google/api_core/bidi.py

```diff
@@ -134,6 +137,69 @@ def __iter__(self):
             yield item
 
 
+class _Throttle(object):
+    """A context manager limiting the total entries in a sliding time window.
+
+    If more than ``entry_cap`` attempts are made to enter the context manager
+    instance in the last ``time window`` seconds, the exceeding requests block
+    until enough time elapses.
+
+    The context manager instances are thread-safe and can be shared between
+    multiple threads. If multiple requests are blocked and waiting to enter,
+    the exact order in which they are allowed to proceed is not determined.
+
+    Example::
+
+        max_three_per_second = Throttle(time_window=1, entry_cap=3)
+
+        for i in range(5):
+            with max_three_per_second as time_waited:
+                print("{}: Waited {} seconds to enter".format(i, time_waited))
+
+    Args:
+        time_window (float): the width of the sliding time window in seconds
+        entry_cap (int): the maximum number of entries allowed in the time window
+    """
+
+    def __init__(self, time_window, entry_cap):
```

Comment:
I am not crazy about the signature. I tried to brainstorm some alternatives. 

Currently: `max_three_per_second = Throttle(time_window=1, entry_cap=3)`

Proposal 1: `max_three_per_second = Throttle(accesses_per_window=3, time_window=1)`
Proposal 2: `max_three_per_second = Throttle(access_limit=3, time_window=1)`

Another thought is does it make sense to have `time_window` take a time delta and not a float representing seconds?

Reply:
I'm indifferent here, whatever looks more readable and easier to use. I personally find the second proposal slightly favorable (shorter).

I'm also fine with the timedelta, more self-explanatory, will change.